feat: support eslint v9

Jungzl opened the pull request:
fix: #92

Jungzl committed:
chore: support eslint v9

robatwilliams commented:
Hi. Thanks for your effort on this. As it's a significant piece of work which we hadn't [already discussed](https://github.com/robatwilliams/es-compat/blob/master/CONTRIBUTING.md#contributing), I need some understanding of the changes made and why.

As I see:
* Dropped Node 16 support
* Upgraded lerna
* Upgraded various other packages (run all, prettier, browserslist ...)
* Changed from npm to pnpm
* No tests against ESLint 8
* ...

Could you elaborate please?

Is it all necessary just to support 9, or is it bundling together some good-to-do general maintenance in the same PR?

Jungzl commented:
Sorry for submitting many changes at once. I believe including excessive irrelevant code solely to accommodate ESLint 9 is somewhat bothersome. It might be better to divide it into several pull requests.
Anyway, I can explain the reasons for these changes to you.
> Dropped Node 16 support

[Node16 LTS is EOL in September 11th, 2023](https://nodejs.org/en/blog/announcements/nodejs16-eol);
[ESLint no longer supports versions below node 18.18](https://eslint.org/docs/latest/use/migrate-to-9.0.0#drop-old-node);

> Upgraded lerna

To be honest, I just upgraded all the dependencies that could be upgraded. I saw #77 before, since lerna actively maintained recently, let's give it a try.

> Upgraded various other packages (run all, prettier, browserslist ...)

I replaced `eslint-plugin-import` with `eslint-plugin-import-x`, the latter is a rewrite of the former with TS and legacy node support dropped.
I replaced `npm-run-all` with [`npm-run-all2`](https://github.com/bcomnes/npm-run-all2?tab=readme-ov-file#npm-run-all2-why), it has documented its advantages.

> Changed from npm to pnpm

pnpm is more suitable for this monorepo, reduce lock files and shadow dependencies

And I didn't test it on ESLint 8 since it's a breaking change, so users who use ESLint 8 should continue to use version 3.2.1.

Jungzl committed:
fix: add packageManager for ci

Jungzl committed:
chore: update root eslint config

Jungzl committed:
test: update tests

Kristinita commented on packages/eslint-plugin-ecmascript-compat-example/eslint.config.js:
```diff
@@ -0,0 +1,20 @@
+import compat from 'eslint-plugin-ecmascript-compat';
```
[**The configuration for eslint-plugin-compat**](https://github.com/amilajack/eslint-plugin-compat#new-config-eslintconfigmjs):

```js
import compat from "eslint-plugin-compat";

```

I think it would be nice to use here a name other than `compat` to prevent conflicts with eslint-plugin-compat.

Thanks.
